Add load balanced sample webservice

danmason opened the pull request:
(#674) The 'hello world' example, with the associated cloudformation file and dockerfile.

Commit:
Add skeleton service with load balanced service

Commit:
Remove private elb

danmason committed:
Add logback.xml to crux-soak

danmason committed:
Split out cloudformation setup

jarohen commented on examples/soak/bin/run_task.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
```
what's this doing? assuming soak's deployed as a service, not a task?

danmason replied:
It is - should kill this, I think we were considering keeping it to manually trigger the ingest task though if we do want a script to run that we should include it in that PR

jarohen commented on examples/soak/setup_ecs_cluster.yml:
```diff
@@ -0,0 +1,242 @@
+        Fn::ImportValue:
+          !Join [':', [!Ref 'VpcStackName', 'VPCId']]
+
+  # Load balancers for getting traffic to containers.
```
I don't mind comments sticking about (I normally wouldn't have comments in CF templates explaining what the AWS services are, assumption is that people will look these up in AWS docs if need be) but let's make sure the comments that do stick about still apply (e.g. we're not using a private load balancer)

jarohen commented on examples/soak/setup_ecs_cluster.yml:
```diff
@@ -0,0 +1,242 @@
+                - 'logs:PutLogEvents'
+              Resource: '*'
+
+  SoakTask:
```
I'd call this `SoakTaskDef` - 'task' could be an ECS task

jarohen commented on examples/soak/setup_ecs_cluster.yml:
```diff
@@ -0,0 +1,242 @@
+AWSTemplateFormatVersion: '2010-09-09'
```
there're a few resources in here that'd be per-application, and a few that are per-cluster - e.g. the load balancer isn't application-specific (the forwarding rule is, though) - let's figure out what'll change together, and that's likely how best to split the CF up too

jarohen commented on examples/soak/setup_vpc.yml:
```diff
@@ -0,0 +1,160 @@
+      SubnetId: !Ref PublicSubnetTwo
+      RouteTableId: !Ref PublicRouteTable
+
+  # Setup networking resources for the private subnets. Containers
```
I suspect, when you do your cost analysis, you might find this costs quite a lot, relative to what it's doing :)

jarohen commented on examples/soak/src/crux/soak/main.clj:
```diff
@@ -0,0 +1,12 @@
+(ns crux.soak.main
+  (:require [ring.util.response :as ring]
```
I'd usually alias this as `resp` - there's plenty of other namespaces in ring

jarohen reviewed (changes requested):
Thanks gents - looks like it's got all the right bits, but not necessarily the right places in CF - let's pair :)

How're you planning to redeploy a new version of the service?

danmason committed:
Remove redundant sections of cloudformation

danmason committed:
Rename usages of `ring` as `resp`

danmason committed:
Reduce Cpu and Memory values of the services

danmason committed:
Remove Nat Gateway from crux-soak cloudformation

danmason committed:
Set `desired-count` on the aws crux-soak service to 0

If set above 0, the creation of the stack will hang indefinitely as it
is waiting for a container image to be added. Will need to add a note
to add an image and then set the desired count within the readme.

danmason committed:
Move `SoakTargetGroup` into setup_ecs cloudformation

The PublicLoadBalancerListener references the SoakTargetGroup as its'
default action, so either the default action needs to be empty or the
SoakTargetGroup needs to be accessible to the Listener.

danmason committed:
Use ENTRYPOINT inplace of CMD in crux-soak Dockerfile.

The Task Definitions use a CMD override of the image, so CMD needed to
be ENTRYPOINT in order for the task to run properly.

danmason committed:
Enable 'AssignPublicIp' for tasks started by the Service

danmason committed:
Add health check grace period to Service.

danmason committed:
Add security group ingress for the application load balancer

jarohen committed:
giving it some space

jarohen committed:
don't need the extra load balancer rule given it's the default route

jarohen committed:
move cf templates to their own directory

jarohen committed:
splitting out ECS cluster from LB

jarohen committed:
add dereg delay to 10s because it's 300s by default, which is crazy long

jarohen committed:
splitting out ECR repo because it's a PITA, can't delete it through CF

jarohen committed:
add name tags where we can because it makes it much easier in the AWS console
